db/migrate/20170823232731_create_preserved_objects.rb

```diff
@@ -0,0 +1,12 @@
+class CreatePreservedObjects < ActiveRecord::Migration[5.1]
+  def change
+    create_table :preserved_objects do |t|
+      t.string :druid, null: false, unique: true, index: true
+      t.integer :version
```

Comment:
When would we expect 'version' to be null or not present? Should this also be a case of 'null: false'?

Reply:
i'm not sure... i had a vague thought along these lines when we were first defining the schema last week.  but what, e.g., if you had a corrupted moab where you couldn't determine the version?  in that case, i think we'd want a catalog entry, but i don't know that we could put a valid integer in this field.  though this is not a use case i'm sure about at all.

Reply:
What would a corrupted Moab look like? Presumably missing version directories, or missing expected files inside one or more version directories. Would we ever even ingest such a Moab into OIS, or violently throw up before we even got to that point? If a known-good, previously-sane Moab later has problems (v7 is fine, but v8 is added and is bad), would we record that Moab as being v7, but with problems, or set version to 'null' and and then handle the exception somehow?

Reply:
That's a round-about way of saying I think requiring a value for version is probably desirable, at the very least as a way of tracking "the last known-good version" of a Moab if we run into subsequent problems.

Reply:
i guess i was imagining something like... we create a top level moab directory for an object, and something dies silently before the first version subdirectory is created.  other processes expect a valid moab object to be there, but while there was a place made for it, a valid moab object was never fully created in the first place.

if this is completely unrealistic, and no similar situation can be imagined, i'd agree that this field should be required.

Reply:
I think in that scenario, either the inventory process or ingest should throw up long before we try to write data to the database. We should take it as a point of pride that we don't ingest bad data and alert very noisily if that happens.

There's also the side-question of "what makes a directory a Moab object?"; we shouldn't pin that to 'the directory is in Moab name format', but rather that the contents of that directory contain everything we'd expect to see in a well-formed Moab, and nothing else.

Reply:
Agreed. We don't have a scenario where we're building up a Moab object from an initially empty directory. It only exists to contain at least one version of the object, so anything short of that should fail to be created.

Reply:
thanks for the info, sounds good, will add the constraint